File: apps/backend-java/src/main/java/com/enterpriseagentos/backendjava/domain/common/primitives/AgentError.java
package com.enterpriseagentos.backendjava.domain.common.primitives;

import java.time.Instant;
import java.util.UUID;
import com.enterpriseagentos.backendjava.domain.features.analytics.AgentLogRecord;
import com.enterpriseagentos.backendjava.domain.features.analytics.AgentLogType;

public record AgentError(AgentErrorCategory category, String message, String detail) {
    public AgentLogType logType() {
        return switch (category) {
            case PodConnection -> AgentLogType.ErrorPodConnection;
            case LlmCall -> AgentLogType.ErrorLlmCall;
            case ToolExecution -> AgentLogType.ErrorToolExecution;
            case SkillExecution -> AgentLogType.ErrorSkillExecution;
            case TurnOrchestration -> AgentLogType.ErrorTurnOrchestration;
            case Memory -> AgentLogType.ErrorMemory;
            case Configuration -> AgentLogType.ErrorConfiguration;
        };
    }

    public String formattedContent() {
        String prefix = category + ": " + message;
        return detail == null ? prefix : prefix + "\n" + detail;
    }

    public AgentLogRecord toLogRecord(UUID agentId) {
        AgentLogRecord record = new AgentLogRecord();
        record.agentId = agentId;
        record.type = logType();
        record.content = formattedContent();
        record.time = Instant.now();
        return record;
    }
}

File: apps/backend/src/Common/Domain/Primitives/AgentError.cs
using OffceOs.Features.ResourceLogs.Domain;

namespace OffceOs.Common.Domain.Primitives;

/// <summary>
/// Structured agent error with category for dashboard filtering.
/// </summary>
public sealed record AgentError(
    AgentErrorCategory Category,
    string Message,
    string? Detail = null)
{
    public ResourceLogType LogType => Category switch
    {
        AgentErrorCategory.PodConnection => ResourceLogType.ErrorPodConnection,
        AgentErrorCategory.LlmCall => ResourceLogType.ErrorLlmCall,
        AgentErrorCategory.ToolExecution => ResourceLogType.ErrorToolExecution,
        AgentErrorCategory.SkillExecution => ResourceLogType.ErrorSkillExecution,
        AgentErrorCategory.TurnOrchestration => ResourceLogType.ErrorTurnOrchestration,
        AgentErrorCategory.Memory => ResourceLogType.ErrorMemory,
        AgentErrorCategory.Configuration => ResourceLogType.ErrorConfiguration,
        _ => ResourceLogType.Error,
    };

    public string FormattedContent => Detail is not null
        ? $"{Category}: {Message}\n{Detail}"
        : $"{Category}: {Message}";

    /// <summary>
    /// Creates an ResourceLogRecord from this error. Use this anywhere outside the turn loop
    /// where TurnLogger isn't available.
    /// </summary>
    public ResourceLogRecord ToLogRecord(Guid agentId) => new()
    {
        AgentId = agentId,
        Type = LogType,
        Content = FormattedContent,
        Time = DateTime.UtcNow,
    };
}
